File: cpp/leetcode/135.cpp
class Solution {
  public:
    int candy(vector<int> &ratings) {
        if (ratings.empty()) {
            return 0;
        }
        int n = ratings.size();
        vector<int> dp(n, 1);
        for (int i = 1; i < n; i++) {
            if (ratings[i] > ratings[i - 1]) {
                dp[i] = dp[i - 1] + 1;
            }
        }
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                dp[i] = max(dp[i], dp[i + 1] + 1);
            }
        }
        int res = 0;
        for (int num : dp) {
            res += num;
        }
        return res;
    }
};

File: python/leetcode/135.py
class Solution(object):
    def candy(self, ratings):
        n = len(ratings)
        res = [1 for _ in xrange(n)]
        for i in xrange(1, n):
            if ratings[i] > ratings[i - 1]:
                res[i] = res[i - 1] + 1
        for i in xrange(n - 1, 0, -1):
            if ratings[i - 1] > ratings[i]:
                res[i - 1] = max(res[i - 1], res[i] + 1)
        total = 0
        for item in res:
            total += item
        return total
